gulpfile.babel.js

```diff
@@ -93,10 +98,80 @@ gulp.task('default', ['clean'], cb =>
     'lint',
     'vulcanize',
     'scripts',
+    'generate-service-worker',
     cb
   )
 );
 
+// Copy over the scripts that are used in importScripts as part of the generate-service-worker task.
+gulp.task('copy-sw-scripts', () => {
+  return gulp.src('node_modules/sw-toolbox/sw-toolbox.js')
+    .pipe(gulp.dest('static/dist'));
+});
+
+// Generate a service worker file that will provide offline functionality for
+// local resources.
+gulp.task('generate-service-worker', ['copy-sw-scripts'], () => {
+  const staticDir = 'static';
+  const distDir = path.join(staticDir, 'dist');
+  const filepath = path.join(distDir, 'service-worker.js');
+
+  // Precache all the truly static resources, and everything else will be
+  // cached at runtime. While the various Polymer elements are static, they are
+  // not pre-cached. Reasons include:
+  //  - The elements may be served in different ways. Typically, the Polymer
+  //    elements are vulcanized into a single file per page, but may be served
+  //    individually during development or testing.
+  //  - The vulcanized files end up being rather large. Rather than consuming
+  //    more space on service worker install, wait until the user visits the
+  //    including page
+  //  - The pages already include logic to preload/push the elements. Rely on
+  //    that logic to improve the initial loading performance of the page.
+  return swPrecache.write(filepath, {
+    cacheId: 'chromestatus',
+    verbose: true,
+    logger: $.util.log,
+    staticFileGlobs: [
+      // Images
+      `${staticDir}/img/**/*`,
+      `${staticDir}/elements/openinnew.svg`,
+      // Scripts
+      `${staticDir}/bower_components/webcomponentsjs/webcomponents-lite.min.js`,
+      `${staticDir}/js/**/*.js`,
+      // Styles
+      `${staticDir}/css/**/*.css`,
+    ],
+    runtimeCaching: [
+      // Server-side generated content
+      {
+        urlPattern: /\/features\/\w+$/,
+        handler: 'fastest',
+        options: {
+          cache: {
+            maxEntries: 10,
+            name: 'features-cache'
+          }
+        }
+      },
+      // For dynamic data (json), try the network first to get the most recent
+      // values.
+      {
+        urlPattern: /\/features.json$/,
+        handler: 'networkFirst',
+      },
+      {
+        urlPattern: /\/omaha_data$/,
+        handler: 'networkFirst',
+      },
+      // Polymer elements that may or may not be vulcanized
+      {
+        urlPattern: /\/static\/(bower_components|elements)\//,
+        handler: 'cacheFirst',
```

Comment:
The problem with using a runtime strategy like `fastest` is that each and every time you request a resource, you're going to fire off an additional network request, and that will end up using bandwidth. (That network request may or may not be fulfilled by the HTTP cache, depending on the standard caching headers.)

Using the precaching and versioning strategy employed by `sw-precache`, you pay the upfront cost once, and then don't fire off any additional network requests unless the service worker detects that there's actually a change. So it ends up being more bandwidth-conscious over time.

Removing unlikely-to-be-needed bundles, like the admin bundle, definitely make sense. And as with any caching strategy, the more fine-grained the resources, the less likely that a large bundle will need to be invalidated due to a small change.